sophia/src/term/iri.rs

```diff
@@ -0,0 +1,854 @@
+    /// IRI is returned else the IRI is returned unchanged.
+    pub fn no_suffix(&self) -> Cow<'_, Self>
+    where
+        TD: From<String>,
+    {
+        match &self.suffix {
+            Some(s) => {
+                let mut full = String::with_capacity(self.len());
+                full.push_str(self.ns.as_ref());
+                full.push_str(s.as_ref());
+                // Okay as derived from existing IRI.
+                let iri = Self::new_unchecked(full, self.absolute);
+                Cow::Owned(iri)
+            }
+            None => Cow::Borrowed(self),
+        }
+    }
+
+    /// Separates the IRI into namespace and suffix at the last hash `#` or
+    /// slash `/` in the IRI.
+    ///
+    /// 1. If this already applies the IRI is returned unchanged.
+    /// 1. If the IRI none of the separators contains the no suffix policy is
+    /// applied.
+    /// 1. In every other case a new IRI is allocated and the policy is
+    ///   applied.
+    pub fn suffixed_at_last_hash_or_slash(&self) -> Cow<'_, Self>
+    where
+        TD: From<String>,
+    {
+        const SEPERATORS: &[char] = &['#', '/'];
+
+        let ns = self.ns.as_ref();
+        match &self.suffix {
+            Some(suf) => {
+                let suf = suf.as_ref();
+                if let Some(pos) = suf.rfind(SEPERATORS) {
+                    // case: suffix with separator
+                    let mut new_ns = String::with_capacity(ns.len() + pos + 1);
+                    new_ns.push_str(ns);
+                    new_ns.push_str(&suf[..=pos]);
+                    let iri = Iri {
+                        ns: new_ns.into(),
+                        suffix: Some(suf[pos + 1..].to_string().into()),
+                        absolute: self.absolute,
+                    };
+                    Cow::Owned(iri)
+                } else if ns.ends_with(SEPERATORS) {
+                    // case: ns does end with separator
+                    Cow::Borrowed(self)
+                } else if let Some(pos) = ns.rfind(SEPERATORS) {
+                    // case: ns does not end with separator but contains one
+                    let mut new_suffix = String::with_capacity(ns.len() - pos - 1 + suf.len());
+                    new_suffix.push_str(&ns[pos + 1..]);
+                    new_suffix.push_str(suf);
+                    let iri = Iri {
+                        ns: ns[..=pos].to_string().into(),
+                        suffix: Some(new_suffix.into()),
+                        absolute: self.absolute,
+                    };
+                    Cow::Owned(iri)
+                } else {
+                    // case: neither contains a separator
+                    let new_ns = self.value();
+                    let iri = Iri {
+                        ns: new_ns.into(),
+                        suffix: None,
+                        absolute: self.absolute,
+                    };
+                    Cow::Owned(iri)
```

Comment:
There is a test case `"tag:", "foo"` -> `"tag:foo"` in `sophia::term::test::iri_normalized_last_hash_or_slash.

Is this really the intended behavior? Or should in this case the suffix rather stay untouched?

Reply:
That's a fair question. Actually, hash_or_slash should probably be replaced by gen-delim (see

Reply:
So should it include all `gen-delims`?

```
gen-delims     = ":" / "/" / "?" / "#" / "[" / "]" / "@"
```

or only `/`, `#` and `:`?

Reply:
I think it would make sense to include all gen-delims.

Reply:
Resolved in e02c4ed